src/pu-checksum.c

```diff
@@ -49,3 +50,53 @@ pu_checksum_verify_file(const gchar *filename,
 
     return TRUE;
 }
+
+gboolean
+pu_checksum_verify_raw(const gchar *filename,
+                       goffset offset,
+                       gsize size,
+                       const gchar *checksum,
+                       GChecksumType checksum_type,
+                       GError **error)
+{
+    g_autofree guchar *buffer = NULL;
+    g_autoptr(GChecksum) checksum_obj = NULL;
+
+    g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+    if (!pu_file_read_raw(filename, &buffer, offset, size, NULL, error))
+        return FALSE;
+
+    checksum_obj = g_checksum_new(checksum_type);
+    g_checksum_update(checksum_obj, buffer, size);
+
+    if (!g_str_equal(checksum, g_checksum_get_string(checksum_obj))) {
+        g_set_error(error, PU_ERROR, PU_ERROR_CHECKSUM,
+                    "Given checksum %s of %s at offset %ld and size %ld does not match %s",
+                    checksum, filename, offset, size, g_checksum_get_string(checksum_obj));
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+gchar *
+pu_checksum_new_from_file(const gchar *filename,
+                          goffset offset,
+                          GChecksumType checksum_type,
+                          GError **error)
+{
+    g_autofree guchar *buffer = NULL;
+    g_autoptr(GChecksum) checksum_obj = NULL;
+    gsize bytes_read = 0;
+
+    g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+    if (!pu_file_read_raw(filename, &buffer, offset, -1, &bytes_read, error))
+        return FALSE;
+
+    checksum_obj = g_checksum_new(checksum_type);
```

Comment:
Look into `g_compute_checksum_for_data ()` for checksum generation.

Reply:
Thanks, did not notice this function before. Using it now.